x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLConfigurationSettings.java

```diff
@@ -143,16 +145,43 @@ public class SSLConfigurationSettings {
         TRUST_STORE_TYPE_TEMPLATE
     );
 
-    private static final Function<String, Setting<Optional<String>>> TRUST_RESTRICTIONS_TEMPLATE = key -> new Setting<>(
+    private static final Function<String, Setting<Optional<String>>> TRUST_RESTRICTIONS_PATH_TEMPLATE = key -> new Setting<>(
         key,
         s -> null,
         Optional::ofNullable,
         Property.NodeScope,
         Property.Filtered
     );
-    private static final SslSetting<Optional<String>> TRUST_RESTRICTIONS = SslSetting.setting(
+    private static final SslSetting<Optional<String>> TRUST_RESTRICTIONS_PATH = SslSetting.setting(
         "trust_restrictions.path",
-        TRUST_RESTRICTIONS_TEMPLATE
+        TRUST_RESTRICTIONS_PATH_TEMPLATE
+    );
+
+    public static final Function<String, Setting<List<String>>> TRUST_RESTRICTIONS_X509_FIELDS_TEMPLATE = key -> Setting.listSetting(
```

Comment:
The dotted names make sense to me, just in case we get requests to parse CN from other parts of a cert in the future. For example:
1. subjectName.commonName
2. issuerName.commonName
3. subjectAltName.otherName.commonName
4. issuerAltName.otherName.commonName
5. subjectAltName.directoryName
6. issuerAltName.directoryName

Reply:
> The existing SSL config code mostly uses enums for these sorts of settings (e.g. clientAuth, verificationMode). Is it intentional that this would be different?

The settings infrastructure does not allows you to define a default that is not a list of strings.  The generic listSettings only support a `Setting<List<T>> fallbackSetting`,  which is very different from a `List<T> defaultList` (which is what we would need).   I see there is support in the `SslConfigurationLoader` via the `resolveListSetting` abstraction. However,  I found that abstraction very confusing since IMO it breaks the contract of how to define settings and where the defaults are derived. Rather than deriving the default from the setting itself , it is derived explicitly at run time via a specialized method and the default is defined in[ business class](https://github.com/elastic/elasticsearch/blob/29aea35a9efab3194ef6a33c403b821ebe8bda94/libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslConfigurationLoader.java#L146) not the setting itself. The setting itself must be defined to have a default value of `null` which is arguably wrong since it violates the settings contract. Since the default value must be defined as [null](https://github.com/elastic/elasticsearch/blob/29aea35a9efab3194ef6a33c403b821ebe8bda94/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/SSLConfigurationSettings.java#L230) (when it is not null), you can not simply pass around the setting and expect to get back the correct default. Further, that abstraction is only in 8.x which would make the backporting even more difficult.  I personally found adding the setting to be by far the most complicated and frustrating piece of this entire change mostly due to the abstraction and mismatched expectations of how you define a default and read a default value (it uniquely different from how any other settings work). I chose to better honor the setting contract and define the default the default in the setting. Enums are nice but we need better support from the setting infrastructure to model the default as a list of enums (I could model the values as an enum as a wrapper around the strings, but that is kinda pointless). Long term, I would like to revisit this abstractions and make the other settings that default as non strings better support defaults as part of the setting definition itself.

Reply:
The code in `ssl-config` is complicated by the fact that libs can't depend on `Setting` (the Setting class is in core, and we can't have a dependency from ssl-config back to core).
So all the "Settings" in ssl-config have to be implemented in a very manual way which includes having to handle default values with explicit code instead of a default value. Technically it would have been possible to workaround it in different ways (essentially by duplicating half of the`Setting` class into ssl-config) but we made the choice to just handle null/empty values in `SslConfigurationLoader`. None of that has anything to do with enums - the code works the same way for string and enums, it's just a consequence of the fact that `Setting` & `Settings` have not been extracted into a standalone library.

`Setting` always treats default values as `String` so that things like `_cluster/settings?include_defaults=true` can work (which it doesn't for SSL settings because we can't use `Setting`, but the SSL settings are filtered anyway, so it wouldn't show anything even if we did register a default). The idea is that every default value for a setting can be represented in the same format that an admin would use to define that setting. So, even if the setting is an integer (see [`intSetting`](https://github.com/elastic/elasticsearch/blob/0652a5915586abbcdfdfffa91a391740f0147752/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L1308)) the default value is a `String` which will then be parsed back into an `Integer` when read.

That doesn't prevent using an enum for a `Setting` type - see for example [`LdapUserSearchSessionFactorySettings.SEARCH_SCOPE`](https://github.com/elastic/elasticsearch/blob/255bf5056bdbae9cd594f7c3e965b96d33087a39/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/ldap/LdapUserSearchSessionFactorySettings.java#L45-L54) or [`LicenseService.ALLOWED_LICENSE_TYPES_SETTING`](https://github.com/elastic/elasticsearch/blob/d1faa903d25fde8eded93e5e84e41fdacdb0caf8/x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicenseService.java#L81-L87) it just means that the default values need to be converted into a `String` (or `List<String>`).  
The benefit is that all the code that checks whether a particular x509_field is configured deals with realm enum objects rather than strings pretending to be enums - all the case conversion / parsing code lives in the `Setting` and anything that uses the Setting value can do a simple `contains` (or for single-value settings, `switch`).

Reply:
I have updated the code to better conform to the existing settings and use an enumeration.  I had to exposed the default value as a final static public variable used in both the buisness class and the setting itself which alieviates my concerns about the differences in the default w.r.t. to the contract. (FWIW, the settings code requires a non-null default value for list settings). Change is over at https://github.com/elastic/elasticsearch/pull/91983.